proposals/readonly-ref.md

```diff
@@ -0,0 +1,470 @@
+        {
+            Test(v);
+        }
+
+        static void Test(in Vector3 v1)
+        {
+            Debug.Assert(v1 == Vector3.UnitY);
+            // changes v1 deterministically (no races required)
+            ChangeV();
+            Debug.Assert(v1 == Vector3.UnitX);
+        }
+
+        static void ChangeV()
+        {
+            v = Vector3.UnitX;
+        }
+```
+
+## Conversions at the call site.
+
+Aliasing indirectly affects what conversions are allowed at the call-site. 
+While explicit and identity conversions are ok, it is not clear whether implicit conversions are ok since they only can be allowed via implicit copying. 
+For starters, let's assume implicit conversions are _not_ ok, but may consider alternatives.
+
+Not allowing implicit conversions would at least be consistent with `out` parameters, although less intuitive since `in` is not spelled out at the call site. It would also be simpler in terms of impact on overload resolution and betterness analysis.
+
+## Invoking instance struct methods on `in` parameters.
+Since all regular instance methods of a struct can potentially mutate the instance or ref-expose `this`, an intermediate copy must be created, as already a case when receiver is a readonly field.
+
+However, since there is no backward compatibility considerations and there are workarounds, compiler should give a warning to ensure the implicit copying is noted by the user.
+
+## Metadata representaion.
+
+CLR provides several mechanisms for augmenting signatures - custom attributes, pseudo custom attributes and custom modifiers (modreq/modopt).   
+Either would work, in theory, but we need to decide on one or another.
+
+Historically `out` uses pseudo custom attribute `[out]`.
+There is a corresponding `[in]` pseudo custom attribute, that could be used.
+
+Attributes can be more attractive for parameters if it is a goal to make a change from ordinary `ref`
+parameter to a `readonly ref` one  nonbreaking to existing callers (in binary, no recompile sense).
+
+`modreq` could be more suitable for `readonly ref` returns if we want to block misuse by the downlevel callers.
+
+# Readonly ref returns
+
+"Readonly ref returns" is an augmentation of ref returns that allows returning references to variables without exposing them to modification.
+
+## Motivation
+The motivation for this sub-feature is roughly symmetrical to the reasons for the `in` parameters - avoiding copying, but on the returning side. A method or an indexer of a nontrivial struct type has currently two options -return by reference and be exposed to possible mutations or copy the value.
+
+## Solution
+`readonly ref` _returns_. (a more concise modifier like `in` so far has been elusive for this case).
+
+```C#
+        struct ImmutableArray<T>
+        {
+            private readonly T[] array;
+
+            public readonly ref T RefAt(int i)
+            {
+                // returning a readonly ref 
+                return ref this.r1;
```

Comment:
Same here. I suggest "return readonly ref this.r1" syntax. A bit lengthy but IMO much more clearer to the reader, and more consistent with "ref" and "out" features.